typescript.js

```diff
@@ -2,79 +2,60 @@ module.exports = {
-            allow: [
-                'arrowFunctions',
-                'functions',
-                'methods'
-            ]
-        }],
-        'no-extra-semi': 'off',
-        '@typescript-eslint/no-extra-semi': 'error',
-        'no-loop-func': 'off',
-        '@typescript-eslint/no-loop-func': 'error',
-        'no-redeclare': 'off',
-        '@typescript-eslint/no-redeclare': 'error',
-        'no-shadow': 'off',
-        '@typescript-eslint/no-shadow': 'error',
-        'no-unused-expressions': 'off',
-        '@typescript-eslint/no-unused-expressions': 'error',
-        'no-unused-vars': 'off',
-        '@typescript-eslint/no-unused-vars': ['error', { vars: 'all', args: 'after-used', ignoreRestSiblings: true }],
-        'no-use-before-define': 'off',
-        '@typescript-eslint/no-use-before-define': ['error', { functions: false, classes: true, variables: true }],
-        'no-useless-constructor': 'off',
-        '@typescript-eslint/no-useless-constructor': 'error',
-        quotes: 'off',
-        '@typescript-eslint/quotes': ['error', 'single', { avoidEscape: true }],
-        'no-return-await': 'off',
-        '@typescript-eslint/return-await': 'error',
-        semi: 'off',
-        '@typescript-eslint/semi': 'error',
-        'space-before-function-paren': 'off',
-        '@typescript-eslint/space-before-function-paren': ['error', {
-            anonymous: 'always',
-            named: 'never',
-            asyncArrow: 'always'
-        }],
-        'space-infix-ops': 'off',
-        '@typescript-eslint/space-infix-ops': 'error',
+
+        // None
+
+        /*
+            TypeScript rules outside of the recommended configuration
+        */
+        '@typescript-eslint/array-type': 'error',
+        '@typescript-eslint/ban-tslint-comment': 'error',
+        '@typescript-eslint/class-literal-property-style': 'error',
+        '@typescript-eslint/consistent-indexed-object-style': 'error',
+        '@typescript-eslint/consistent-type-assertions': 'error',
+        '@typescript-eslint/consistent-type-definitions': 'error',
+        '@typescript-eslint/consistent-type-imports': 'error',
+        '@typescript-eslint/explicit-function-return-type': 'error',
+        '@typescript-eslint/explicit-member-accessibility': 'error',
+        '@typescript-eslint/member-delimiter-style': 'error',
+        '@typescript-eslint/member-ordering': 'error',
+        '@typescript-eslint/method-signature-style': 'error',
+        '@typescript-eslint/no-confusing-non-null-assertion': 'error',
+        '@typescript-eslint/no-dynamic-delete': 'error',
+        '@typescript-eslint/no-extraneous-class': 'error',
+        '@typescript-eslint/no-implicit-any-catch': 'error',
+        '@typescript-eslint/no-invalid-void-type': 'error',
+        '@typescript-eslint/no-parameter-properties': 'error',
+        '@typescript-eslint/no-require-imports': 'error',
+        '@typescript-eslint/no-type-alias': 'error',
```

Comment:
[@typescript-eslint/no-type-alias](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-type-alias.md)
The Web Portal codebase has a pattern of using type aliases in unions to form string enums:
```ts
export type HighlightRuleType = 'lt' | 'gt' | 'between';
```
This seems fine to me and we could allow it via `"allowAliases": "in-unions"`

I also saw a handful of uses to make code more readable, especially in ngrx reducers with generics:
```ts
export type DataSourceSchemaState = EntityState<DataSourceSchema>;
```
After reading the docs I don't understand what the problem is with the above code so I'm tempted to disable the rule entirely, but maybe I'm missing something.

Reply:
Types are absolutely useful, and I didn't find their examples convincing. Claiming the feature should be disabled, because there are bad practices isn't a good argument. Examples like primitive unions are common type definitions, so why not allow an alias? Many of the union and intersection examples don't reflect real world cases, e.g ```string & string[]```. Do we really have someone that would write  ```type myString = string```? Alias unions, callbacks, mapped types, and tuples are common. I've actually used constructors for mixins. I do agree that interfaces are preferred over object literals.

**Proposal**
Disallow object literals in favor of interfaces except where explicitly disabled with a comment.

```typescript
/* 
  Type aliases should be used to promote readable and DRY coding practices. However, interfaces should be preferred
  over object literals.
*/
'@typescript-eslint/no-type-alias': ['error', {
  allowAliases: 'always',
  allowCallbacks: 'always',
  allowConditionalTypes: 'always',
  allowConstructors: 'always',
  allowLiterals: 'never',
  allowMappedTypes: 'always',
  allowTupleTypes: 'always'
}]
```

Reply:
I'll double check that Trevor's proposed configuration would allow the cases I described in my original comment but otherwise there doesn't seem to be any objection to Trevor's proposal.

Reply:
In turns out, interfaces are already enforced by [@typescript-eslint/consistent-type-definitions](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-definitions.md), so I have a new proposal!

**Proposal**
Leave this rule unconfigured.

Reply:
@rajsite This is awaiting your vote

Reply:
Having `'@typescript-eslint/no-type-alias': 'off'` makes sense. The docs are not clear what the default configuration for `@typescript-eslint/consistent-type-definitions` is.

Does the proposal imply requiring interfaces so a configuration of `"@typescript-eslint/consistent-type-definitions": ["error", "interface"]`?

Reply:
Yes, I guess I figured we would get to it, but it doesn't look like it's in the recommended set.